File: examples/paper_example03.jl

using OptimalTransportNetworks
using Random
using Plots

# ------------------------------------
# 4.2 Random Cities with Multiple Goods
# -------------------------------------

ngoods = 10
width = 9  
height = 9

# Init and Solve network

param = init_parameters(K = 10000, labor_mobility = true, annealing = false)
graph = create_graph(param, width, height, type = "triangle")

# set fundamentals

param[:N] = 1 + ngoods
graph[:Zjn] = zeros(graph[:J], param[:N]) # matrix of productivity

graph[:Zjn][:,1] .= 1 # the entire countryside produces the homogenenous good 1

if ngoods > 0
    Ni = find_node(graph, 5, 5) # center
    graph[:Zjn][Ni, 2] = 1 # central node
    graph[:Zjn][Ni, 1] = 0 # central node
    
    Random.seed!(5)
    for i in 2:ngoods
        newdraw = false
        while newdraw == false
            j = round(Int, 1 + rand() * (graph[:J] - 1))
            if graph[:Zjn][j, 1] > 0
                newdraw = true
                graph[:Zjn][j, i+1] = 1
                graph[:Zjn][j, 1] = 0
            end
        end
    end
end

# Convex case
results = Array{Any}(undef, 2)
results[1] = optimal_network(param, graph)

# Nonconvex
param[:gamma] = 2
results[2] = optimal_network(graph, I0 = results[1][:Ijk])

# Plot results

cols = 3 # number of columns
rows = Int(ceil((1 + param[:N]) / cols))

for j in 1:2
    # Initialize an empty array to hold the subplots
    plots = Vector{Any}(undef, (1 + param[:N])) 
    # Plot network
    plots[1] = plot_graph(graph, results[j][:Ijk], node_shades = results[j][:Lj], node_sizes = results[j][:Lj], node_sizes_scale = 40)
    title!(plots[1], "(a) Transport Network")
    # Plot goods flows
    for i in 1:param[:N]
        plots[i+1] = plot_graph(graph, results[j][:Qjkn][:, :, i], edge_color = :brown, arrows = true, arrow_style = "thin", 
                                node_sizes = results[j][:Yjn][:, i], node_sizes_scale = 40,
                                node_shades = graph[:Zjn][:, i])
        title!(plots[i+1], string('(', Char(96 + i + 1), ')', " Flows Good ", i))
    end
    # Combine plots
    final_plot = plot(plots..., layout = (cols, rows), size = (rows*400, cols*400))
    display(final_plot)
end

File: OptimalTransportNetworkToolbox_v.1.0.4b/Code/paper_example03.m
%{
==============================================================
 OPTIMAL TRANSPORT NETWORKS IN SPATIAL EQUILIBRIUM
 by P. Fajgelbaum, E. Schaal, D. Henricot, C. Mantovani 2017-19
 ================================================ version 1.0.4

paper_example03.m: script that generate illustrative example 03 from the
paper.

-----------------------------------------------------------------------------------
REFERENCE: "Optimal Transport Networks in Spatial Equilibrium" (2019) by Pablo D.
Fajgelbaum and Edouard Schaal.

Copyright (c) 2017-2018, Pablo D. Fajgelbaum, Edouard Schaal
pfajgelbaum@ucla.edu, eschaal@crei.cat

This code is distributed under BSD-3 License. See LICENSE.txt for more information.
-----------------------------------------------------------------------------------
%}

clear;
close all;

%% ------------------------------------
% 4.2 Random Cities with Multiple Goods
% -------------------------------------

ngoods=10;
width=9; height=9;

% Init and Solve network

param = init_parameters('TolKappa',1e-5,'K',10000,'LaborMobility','on','Annealing','off');
[param,g]=create_graph(param,width,height,'Type','triangle');

% set fundamentals

param.N = 1+ngoods;
param.Zjn = 0*ones(g.J,param.N); % matrix of productivity

param.Zjn(:,1)=1; % the entire countryside produces the homogenenous good 1

if ngoods>0
    Ni=find_node(g,5,5); % center
    param.Zjn(Ni,2)=1; % central node
    param.Zjn(Ni,1)=0; % central node
    
    rng(5);
    for i=2:ngoods
        newdraw=false;
        while newdraw==false
            j=round(1+rand()*(g.J-1));
            if param.Zjn(j,1)>0
                newdraw=true;
                param.Zjn(j,i+1)=1;
                param.Zjn(j,1)=0;
            end
        end
    end
end

% Convex case
results(1)=optimal_network(param,g);

% Nonconvex
param.gamma=2;
results(2)=optimal_network(param,g,results(1).Ijk);


%% Plot results
close all;

cols = 3; %4 % number of columns
rows = ceil((1+param.N)/cols);

s={'random_cities_multigoods_convex','random_cities_multigoods_nonconvex'};

for j=1:2
    fig=figure('Units','inches','Position',[0,0,7.5,11],'Name',char(s(j)));

    % Plot network
    subplot(rows,cols,1);
    plot_graph( param,g,results(j).Ijk, 'Shades', (results(j).Lj-min(results(j).Lj))./(max(results(j).Lj)-min(results(j).Lj)), 'Sizes', 1+16*(results(j).Lj./mean(results(j).Lj)-1), 'NodeFgColor',[.6 .8 1],'Transparency','off' );
    h=title('(a) Transport Network (I_{jk})','FontWeight','normal','FontName','Times','FontSize',9);

    for i=1:param.N
        subplot(rows,cols,i+1);
        sizes=3*results(j).Yjn(:,i)./sum(results(j).Yjn(:,i));
        shades=param.Zjn(:,i)./max(param.Zjn(:,i));
        plot_graph( param,g,results(j).Qjkn(:,:,i),'Arrows','on','ArrowScale',1,'ArrowStyle','thin','Nodes','on','Sizes',sizes,'Shades',shades, 'NodeFgColor',[.6 .8 1],'Transparency','off' );
        title(sprintf('(%c) Shipping (Q^{%d}_{jk})',97+i,i),'FontWeight','normal','FontName','Times','FontSize',9);
    end

    % Save
%     print('-depsc2',[graph_path,char(s(j)),'.eps'],'-r400');
%     saveas(fig,[graph_path,char(s(j)),'.jpg']);
end

